lib/internal/util/debuglog.js

```diff
@@ -58,16 +60,17 @@ function debuglogImpl(set) {
 // so it needs to be called lazily in top scopes of internal modules
 // that may be loaded before these run time states are allowed to
 // be accessed.
-function debuglog(set) {
-  let debug;
-  return function(...args) {
-    if (debug === undefined) {
-      // Only invokes debuglogImpl() when the debug function is
-      // called for the first time.
-      debug = debuglogImpl(set);
-    }
-    if (debug !== null)
-      debug(...args);
+function debuglog(set, cb) {
+  let debug = (...args) => {
+    // Only invokes debuglogImpl() when the debug function is
+    // called for the first time.
+    debug = debuglogImpl(set);
+    if (typeof cb === 'function')
+      cb(debug);
+    debug(...args);
+  };
+  return (...args) => {
+    debug(...args);
   };
```

Comment:
What about this?

```suggestion
function debuglog(set) {
  let debug = (...args) => {
    // Only invokes debuglogImpl() when the debug function is
    // called for the first time.
    debug = debuglogImpl(set);
    debug(...args);
  };
  const state = { called: false };
  return ((...args) => {
    if(!this.called) {
      debug(...args);
      this.called = true;
    } else {
      debug(...args);
    }
  }).bind(state);
```

just like

https://github.com/nodejs/node/blob/ec204d86b06e4cc9259c1308e365a3e104212a16/lib/events.js#L426-L432

Reply:
because I don't think `let debug = fn('xxx', v => debug = v)` is a good code style

Reply:
The suggested change is basically what the code was doing before. Having a callback is the only way to be able to avoid the wrapper function and utilizing it provides a further, noticeable performance improvement.

Reply:
Thanks, I understand

Reply:
Fairly certain this is because of the way V8 optimizes but that's only a guess. It's interesting that it's such a big perf difference.